### PR TITLE
feat: #686 External API for FSP Tracker

### LINKED ISSUE #686: New FOM API for FSP TS
**Describe the task**
New API.
The API will be public accessible and allow to search for FOM related to (by) FSP ID. Apply naming convention to designate this API as used by external parties.

**Acceptance Criteria**
- [x] New API is built with search params available (FSP ID)
- [x] Add Auth guard if needed.
- [x] Add/Create/Adjust tests as needed.

**Additional context**
- Design discussion around what the API returns: existing `ProjectPublicSummaryResponse` as result, or with new response object for limited information.

[FSP TS FOM API Request.pdf](https://github.com/user-attachments/files/16732139/FSP.TS.FOM.API.Request.pdf)

Image

### PR DESCRIPTION
**Pull request by ianliuwk1019**
Closes #686 
- New controller and module for new external fsp-tracker api.

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-38.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-38.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-38.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

File: api/src/app/modules/external/fsp-tracker/fsp-tracker.controller.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@Controller("external")
+export class FspTrackerController {
+  constructor(
+    private readonly service: FspTrackerService,
+    private readonly _logger: PinoLogger) {
+  }
+
+  @Get("fsp-tracker")
+  @ApiQuery({ name: 'fspId', required: true})
+  @ApiResponse({ status: HttpStatus.OK, type: [FomFspTrackerResponse] })
```
Comment (ianliuwk1019) [inline api/src/app/modules/external/fsp-tracker/fsp-tracker.controller.ts]: Hi @basilv as discussed, this API is separated into different controller.
I intentionally make it into a folder structure like this: "api/src/app/modules/**external**/**fsp-tracker**"; and placed the "**fsp-tracker module**" as a submodule under "**external**" module.
And the endpoint path also look like this (GET api/**external**/**fsp-tracker**):
image

Welcome to suggest different name or if you don't feel "**external**" folder/module is needed then we can change.
(For "external", the intention was for all other external services, like FSP Tracker, to interface with FOM be under a common module)
Thank you!

Comment (basilv) [reply api/src/app/modules/external/fsp-tracker/fsp-tracker.controller.ts]: I like the /external folder, but I think /fsp-tracker should be changed to something like /projects-by-fsp

Comment (ianliuwk1019) [reply api/src/app/modules/external/fsp-tracker/fsp-tracker.controller.ts]: Hi Basil, for folder structure, I will use "modules/external/projects-by-fsp", but for endpoint path, the FSP Tracker team probably would be confused by what "**project**" means... would it be using "**fom**" at endpoint path "GET **api/external/fom-by-fsp**" or ""GET **api/external/fom** (query parameter is on **fspId** already)" make more sense to them?

Comment (basilv) [reply api/src/app/modules/external/fsp-tracker/fsp-tracker.controller.ts]: Hmm, yes /fom makes more sense than project.

File: api/src/core/utils.ts [hunk 2]
```diff
@@ -66,3 +66,8 @@ export const flatDeep = (arr, d = 1) => {
   return d > 0 ? arr.reduce((acc, val) => acc.concat(Array.isArray(val) ? flatDeep(val, d - 1) : val), [])
                 : arr.slice();
 };
+
+// Partial type with nested properties also as Partial.
+export type RecursivePartial<T> = {
+    [P in keyof T]?: RecursivePartial<T[P]>;
+};
```
Comment (basilv) [inline api/src/core/utils.ts:73]: Why is this needed?

**Review by basilv — APPROVED** [review]
One question but looks good otherwise.

**Comment by basilv** [comment]
The response has a little more detail (fom name, client name) than strictly necessary but that's fine and may be helpful for either troubleshooting or to accomodate requirement expansion by fsp-tracker.

File: api/src/core/utils.ts (hunk 2 above)
Comment (ianliuwk1019) [reply api/src/core/utils.ts:73]: In the new test that when supplying the mock data with type `Project` entity, I don't want to code many of unnecessary fields (just id, name and forestClient), so I used typescript `Partial<Project>`. However, this is not in deep level, so it still had error for the type for something like `project.forestClient` relationship field when building a mock data. So, this piece will make it a deep `Partial`.

Comment (basilv) [reply api/src/core/utils.ts:73]: Sounds good.
